apps/omg_child_chain/lib/omg_child_chain/block_queue/submission_monitor.ex

```diff
@@ -0,0 +1,174 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.ChildChain.BlockQueue.SubmissionMonitor do
+  @moduledoc """
+  Listens to block events and raises :block_submit_stalled alarm when a pending block
```

Comment:
so essentially what you want to figure out is:
the difference between blocks mined minus blocks formed, correct?

If yes, why are you storing pending blocks?

Reply:
Mainly to track `{blknum, root_chain_height}`. There was a lot of false alarms when I tracked just the `mined - formed` because of the large swing in block time. And it's a large swing all the time, not just once in a while, so the alarm is barely useful with an alarm every few minutes.

Some supporting info. Below is rootchain blocktime over a single day. You can see the spikes throughout the day, multiple times each hour:

![blocktime](https://user-images.githubusercontent.com/921194/90485220-73da8700-e161-11ea-9070-3846b156e705.png)

And in the diagram below, if we use the average blocktime of 12.5 - 15s as usual, only 6% (369/6401) of all blocks in a day will fall in the range.

![Histogram of blocktime](https://user-images.githubusercontent.com/921194/90485339-9a98bd80-e161-11ea-81d8-5fc7e3f0f6a0.png)

So in my trial it was impossible to stick to `mined - formed`, because over a single day `mined - formed >= 4` could be from 10 seconds to 188 seconds. So tracking the chch blknum against the first root chain height it was submitted at was the most reliable way to track stalled submissions as far as I tried.

---

Source data: https://docs.google.com/spreadsheets/d/1wzniBoO2nonRo83im0VRcaSPj_pr-ZZXKV-yiOOFVOo/edit